README.md
# leaderboard

Leaderboards backed by [Redis](http://redis.io) in Python.

Builds off ideas proposed in http://www.agoragames.com/blog/2011/01/01/creating-high-score-tables-leaderboards-using-redis/.

## Installation

`pip install leaderboard`

Make sure your redis server is running! Redis configuration is outside the scope of this README, but
check out the [Redis documentation](http://redis.io/documentation).

## Usage

### Creating a leaderboard

Be sure to require the leaderboard library:

```python
from leaderboard.leaderboard import Leaderboard
```

Create a new leaderboard or attach to an existing leaderboard named 'highscores':

```python
highscore_lb = Leaderboard('highscores')
```

### Defining leaderboard options

The default options are as follows:

```python
    DEFAULT_PAGE_SIZE = 25
    DEFAULT_REDIS_HOST = 'localhost'
    DEFAULT_REDIS_PORT = 6379
    DEFAULT_REDIS_DB = 0
    DEFAULT_MEMBER_DATA_NAMESPACE = 'member_data'
    DEFAULT_GLOBAL_MEMBER_DATA = False
    ASC = 'asc'
    DESC = 'desc'
    MEMBER_KEY = 'member'
    MEMBER_DATA_KEY = 'member_data'
    SCORE_KEY = 'score'
    RANK_KEY = 'rank'
    DEFAULT_CLUSTER_MODE = False
```

You would use the option, `order=Leaderboard.ASC`, if you wanted a leaderboard sorted from lowest-to-highest score. You may also set the `order` option on a leaderboard after you have created a new instance of a leaderboard. The various `..._KEY` options above control what data is returned in the hash of leaderboard data from calls such as `leaders` or `around_me`. Finally, the `global_member_data` option allows you to control whether optional member data is per-leaderboard (`False`) or global (`True`).

### Using cluster mode

```python
Leaderboard(
    <leader_board_name>,
    host=<cluster_endpoint>,
    db=0, # only one db supported in cluster mode
    cluster_mode=True
)
```

### Ranking members in the leaderboard

Add members to your leaderboard using `rank_member`:

```python
for index in range(1, 11):
  highscore_lb.rank_member('member_%s' % index, index)
```

You can call `rank_member` with the same member and the leaderboard will be updated automatically.

Get some information about your leaderboard:

```python
highscore_lb.total_members()
10

highscore_lb.total_pages()
1
```

Get some information about a specific member(s) in the leaderboard:

```python
highscore_lb.score_for('member_4')
4.0

highscore_lb.rank_for('member_4')
7

highscore_lb.rank_for('member_10')
1
```

### Retrieving members from the leaderboard

Get page 1 in the leaderboard:

```python
highscore_lb.leaders(1)

[{'member': 'member_10', 'score': 10.0, 'rank': 1}, {'member': 'member_9', 'score': 9.0, 'rank': 2}, {'member': 'member_8', 'score': 8.0, 'rank': 3}, {'member': 'member_7', 'score': 7.0, 'rank': 4}, {'member': 'member_6', 'score': 6.0, 'rank': 5}, {'member': 'member_5', 'score': 5.0, 'rank': 6}, {'member': 'member_4', 'score': 4.0, 'rank': 7}, {'member': 'member_3', 'score': 3.0, 'rank': 8}, {'member': 'member_2', 'score': 2.0, 'rank': 9}, {'member': 'member_1', 'score': 1.0, 'rank': 10}]
```

Add more members to your leaderboard:

```python
for index in range(50, 96):
  highscore_lb.rank_member('member_%s' % index, index)

highscore_lb.total_pages()
3
```

Get an "Around Me" leaderboard page for a given member, which pulls members above and below the given member:

```python
highscore_lb.around_me('member_53')

[{'member': 'member_65', 'score': 65.0, 'rank': 31}, {'member': 'member_64', 'score': 64.0, 'rank': 32}, {'member': 'member_63', 'score': 63.0, 'rank': 33}, {'member': 'member_62', 'score': 62.0, 'rank': 34}, {'member': 'member_61', 'score': 61.0, 'rank': 35}, {'member': 'member_60', 'score': 60.0, 'rank': 36}, {'member': 'member_59', 'score': 59.0, 'rank': 37}, {'member': 'member_58', 'score': 58.0, 'rank': 38}, {'member': 'member_57', 'score': 57.0, 'rank': 39}, {'member': 'member_56', 'score': 56.0, 'rank': 40}, {'member': 'member_55', 'score': 55.0, 'rank': 41}, {'member': 'member_54', 'score': 54.0, 'rank': 42}, {'member': 'member_53', 'score': 53.0, 'rank': 43}, {'member': 'member_52', 'score': 52.0, 'rank': 44}, {'member': 'member_51', 'score': 51.0, 'rank': 45}, {'member': 'member_50', 'score': 50.0, 'rank': 46}, {'member': 'member_10', 'score': 10.0, 'rank': 47}, {'member': 'member_9', 'score': 9.0, 'rank': 48}, {'member': 'member_8', 'score': 8.0, 'rank': 49}, {'member': 'member_7', 'score': 7.0, 'rank': 50}, {'member': 'member_6', 'score': 6.0, 'rank': 51}, {'member': 'member_5', 'score': 5.0, 'rank': 52}, {'member': 'member_4', 'score': 4.0, 'rank': 53}, {'member': 'member_3', 'score': 3.0, 'rank': 54}, {'member': 'member_2', 'score': 2.0, 'rank': 55}]
```

Get rank and score for an arbitrary list of members (e.g. friends) from the leaderboard:

```python
highscore_lb.ranked_in_list(['member_1', 'member_62', 'member_67'])

[{'member': 'member_1', 'score': 1.0, 'rank': 56}, {'member': 'member_62', 'score': 62.0, 'rank': 34}, {'member': 'member_67', 'score': 67.0, 'rank': 29}]
```

Retrieve members from the leaderboard in a given score range:

```python
highscore_lb.members_from_score_range(4, 19)

[{'member': 'member_10', 'score': 10.0, 'rank': 47}, {'member': 'member_9', 'score': 9.0, 'rank': 48}, {'member': 'member_8', 'score': 8.0, 'rank': 49}, {'member': 'member_7', 'score': 7.0, 'rank': 50}, {'member': 'member_6', 'score': 6.0, 'rank': 51}, {'member': 'member_5', 'score': 5.0, 'rank': 52}, {'member': 'member_4', 'score': 4.0, 'rank': 53}]
```

Retrieve a single member from the leaderboard at a given position:

```python
highscore_lb.member_at(4)

{'member': 'member_92', 'score': 92.0, 'rank': 4}
```

Retrieve a range of members from the leaderboard within a given rank range:

```python
highscore_lb.members_from_rank_range(1, 5)

[{'member': 'member_95', 'score': 95.0, 'rank': 1}, {'member': 'member_94', 'score': 94.0, 'rank': 2}, {'member': 'member_93', 'score': 93.0, 'rank': 3}, {'member': 'member_92', 'score': 92.0, 'rank': 4}, {'member': 'member_91', 'score': 91.0, 'rank': 5}]
```

#### Optional member data notes

If you use optional member data, the use of the `remove_members_in_score_range` or `remove_members_outside_rank` methods
will leave data around in the member data hash. This is because the internal Redis method, `zremrangebyscore`,
only returns the number of items removed. It does not return the members that it removed.

#### Leaderboard request options

You can pass various options to the calls `leaders`, `all_leaders`, `around_me`, `members_from_score_range`, `members_from_rank_range` and `ranked_in_list`. Valid options are:

* `with_member_data` - `true` or `false` to return the optional member data.
* `page_size` - An integer value to change the page size for that call.
* `members_only` - `true` or `false` to return only the members without their score and rank.
* `sort_by` - Valid values for `sort_by` are `score` and `rank`.

### Conditionally rank a member in the leaderboard

You can pass a function to the `rank_member_if` method to conditionally rank a member in the leaderboard. The function is passed the following 5 parameters:

* `member`: Member name.
* `current_score`: Current score for the member in the leaderboard. May be `nil` if the member is not currently ranked in the leaderboard.
* `score`: Member score.
* `member_data`: Optional member data.
* `leaderboard_options`: Leaderboard options, e.g. 'reverse': Value of reverse option

```python
def highscore_check(self, member, current_score, score, member_data, leaderboard_options):
  if (current_score is None):
    return True
  if (score > current_score):
    return True
  return False

highscore_lb.rank_member_if(highscore_check, 'david', 1337)
highscore_lb.score_for('david')

1337.0

highscore_lb.rank_member_if(highscore_check, 'david', 1336)
highscore_lb.score_for('david')

1337.0

highscore_lb.rank_member_if(highscore_check, 'david', 1338)
highscore_lb.score_for('david')

1338.0
```

### Ranking a member across multiple leaderboards

```python
highscore_lb.rank_member_across(['highscores', 'more_highscores'], 'david', 50000, { 'member_name': 'david' })
```

### Alternate leaderboard types

The leaderboard library offers 3 styles of ranking. This is only an issue for members with the same score in a leaderboard.

Default: The `Leaderboard` class uses the default Redis sorted set ordering, whereby different members having the same score are ordered lexicographically. As per the Redis documentation on Redis sorted sets, "The lexicographic ordering used is binary, it compares strings as array of bytes."

Tie ranking: The `TieRankingLeaderboard` subclass of `Leaderboard` allows you to define a leaderboard where members with the same score are given the same rank. For example, members in a leaderboard with the associated scores would have the ranks of:

```
| member     | score | rank |
-----------------------------
| member_1   | 50    | 1    |
| member_2   | 50    | 1    |
| member_3   | 30    | 2    |
| member_4   | 30    | 2    |
| member_5   | 10    | 3    |
```

The `TieRankingLeaderboard` accepts one additional option, `ties_namespace` (default: ties), when initializing a new instance of this class. Please note that in its current implementation, the `TieRankingLeaderboard` class uses an additional sorted set to rank the scores, so please keep this in mind when you are doing any capacity planning for Redis with respect to memory usage.

Competition ranking: The `CompetitionRankingLeaderboard` subclass of `Leaderboard` allows you to define a leaderboard where members with the same score will have the same rank, and then a gap is left in the ranking numbers. For example, members in a leaderboard with the associated scores would have the ranks of:

```
| member     | score | rank |
-----------------------------
| member_1   | 50    | 1    |
| member_2   | 50    | 1    |
| member_3   | 30    | 3    |
| member_4   | 30    | 3    |
| member_5   | 10    | 5    |
```

## Performance Metrics

You can view [performance metrics](https://github.com/agoragames/leaderboard#performance-metrics) for the
leaderboard library at the original Ruby library's page.

## Ports

The following ports have been made of the [leaderboard gem](https://github.com/agoragames/leaderboard).

Officially supported:

* CoffeeScript: https://github.com/agoragames/leaderboard-coffeescript
* Python: https://github.com/agoragames/leaderboard-python
* Ruby: https://github.com/agoragames/leaderboard

Unofficially supported (they need some feature parity love):

* Java: https://github.com/agoragames/java-leaderboard
* PHP: https://github.com/agoragames/php-leaderboard
* Scala: https://github.com/agoragames/scala-leaderboard

## Contributing to leaderboard

* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
* Fork the project
* Start a feature/bugfix branch
* Commit and push until you are happy with your contribution
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the version or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

## Copyright

Copyright (c) 2011-2018 Ola Mork, David Czarnecki. See LICENSE.txt for further details.



CHANGELOG.md
# CHANGELOG

## 3.7.3 (2018-05-04)

* Fix an edge case in the updated comparison for non-members

## 3.7.2 (2018-05-03)

* Fix running under Python 3 for `ranked_in_list(...)` if `include_missing` is `True` and non-member(s) passed in

## 3.7.1 (2018-02-20)

* Fix `around_me_in` for Python 3+ [#49](https://github.com/agoragames/leaderboard-python/issues/49)
* Fix TieRankingLeaderboard init [#47](https://github.com/agoragames/leaderboard-python/pull/51)

## 3.7.0 (2017-02-03)

* Fixed an error with the `ranked_in_list` method when using certain options [#46](https://github.com/agoragames/leaderboard-python/pull/46)
* Added `total_scores` method [#45](https://github.com/agoragames/leaderboard-python/pull/45)

## 3.6.1 (2016-10-18)

* Remove `izip_longest` import where unneeded. [#44](https://github.com/agoragames/leaderboard-python/pull/44)

## 3.6.0 (2016-09-16)

* More pipelining as per Ruby library, [#58](https://github.com/agoragames/leaderboard/pull/58).
  Also adds `members_data_for(...) and `members_data_for_in(...)` methods.
* Guard against an out of bounds index in `member_at`

## 3.5.0 (2015-12-11)

* Allow options to be passed down to the connection pool.

## 3.4.0 (2015-06-16)

* Allow a Redis connection to be passed in the Leaderboard initializer using the `redis_connection` option.

## 3.3.0 (2015-03-20)

* Similar fix to one found in the [Ruby leaderboard](https://github.com/agoragames/leaderboard) library. Fixes TieRankingLeaderboard doesn't rank if the score is 0.
* Allow member data to be set in the `change_score_for(...)` method.
* Add `include_missing` option in leaderboard request options to change
  whether or not to include missing members in the result.

## 3.2.0 (2015-02-15)

* Add `global_member_data` option that allows multiple leaderboards to share the same set of member_data. [#51](https://github.com/agoragames/leaderboard/pull/51) for original pull request from the Ruby leaderboard library.
* Add `top` helper method. [#50](https://github.com/agoragames/leaderboard/pull/50) for the original pull request from the Ruby leaderboard library.

## 3.1.0 (2014-11-07)

* Add support for `change_score_for(...)` in the `TieRankingLeaderboard` class.

## 3.0.1 (2014-09-04)

* Add Python3 version check to accommodate 2to3 change of itertools.izip_longest [#30](https://github.com/agoragames/leaderboard-python/pull/30)

## 3.0.0 (2014-07-28)

* Add support for tie handling in leaderboards [#28](https://github.com/agoragames/leaderboard-python/pull/28)
* Re-organized the structure of the leaderboard library so that all the base leaderboard code is no longer in the init.py.

## 2.8.0 (2014-02-15)

* Allow for customization of member_data namespace [#19](https://github.com/agoragames/leaderboard-python/pull/19)

## 2.7.0 (2014-01-24)

* Allow for custom keys to be set for customizing the data returned from calls like `leaders` or `around_me` [#18](https://github.com/agoragames/leaderboard-python/pull/18). Thanks @seaders.

## 2.6.1 (2014-01-22)

* Identify client class and send appropriate zadd command. Resolves [#17](https://github.com/agoragames/leaderboard-python/issues/17).

## 2.6.0 (2013-11-12)

* Added `score_for_percentile` method to be able to calculate the score for a given percentile value in the leaderboard.

## 2.5.0 (2013-07-17)

* Added `rank_member_across` method to be able to rank a member across multiple leaderboards at once.
* Fixed bugs in `leaders_in` and `around_me_in` methods that would not correctly use the `leaderboard_name` argument.

## 2.4.0 (2013-05-31)

* Added `remove_members_outside_rank` method to remove members from the leaderboard outside a given rank.

## 2.3.0 (2013-05-15)

* Added `members_only` option for various leaderboard requests - HT: [Simon Zimmerman](https://github.com/simonz05)
* `leaders` call should return `[]` in case of an empty result set - HT: [Simon Zimmerman](https://github.com/simonz05)
* Initializer no longer deep copies options - HT: [Simon Zimmerman](https://github.com/simonz05)

## 2.2.2 (2013-02-22)

* Fixed a data leak in `expire_leaderboard` and `expire_leaderboard_at` to also set expiration on the member data hash.

## 2.2.1 (2012-12-19)

* Updated `remove_member` to also remove the optional member data for the member being removed.

## 2.2.0 (2012-12-03)

* Added `rank_member_if` and `rank_member_if_in` methods that allow you to rank a member in the leaderboard based on execution of a function.
* Added `rank_members` and `rank_members_in` methods that allow you to rank multiple members in a leaderboard at once by passing in an array of members and scores.

## 2.1 (2012-11-27)

* No longer cast scores to a floating point automatically. If requesting a score for an unknown member in the leaderboard, return `None`. Under the old behavior, a `None` score gets returned as 0.0. This is misleading as 0.0 is a valid score.
* Fixes a bug in `ranked_in_list` when requesting a list with an unknown member.

## 2.0.1 (2012-11-15)

* Remove unnecessary options from `score_and_rank_for` method

## 2.0.0 (2012-11-14)

* Port missing functionality from Ruby [leaderboard](https://github.com/agoragames/leaderboard) project to Python.
* Feature and data parity should exist between these two implementations.

test/leaderboard/__init__.py
from __future__ import absolute_import
import unittest
from .leaderboard_test import LeaderboardTest
from .tie_ranking_leaderboard_test import TieRankingLeaderboardTest
from .competition_ranking_leaderboard_test import CompetitionRankingLeaderboardTest
from .reverse_tie_ranking_leaderboard_test import ReverseTieRankingLeaderboardTest
from .reverse_competition_ranking_leaderboard_test import ReverseCompetitionRankingLeaderboardTest


def all_tests():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(LeaderboardTest))
    suite.addTest(unittest.makeSuite(TieRankingLeaderboardTest))
    suite.addTest(unittest.makeSuite(ReverseTieRankingLeaderboardTest))
    suite.addTest(unittest.makeSuite(CompetitionRankingLeaderboardTest))
    suite.addTest(unittest.makeSuite(ReverseCompetitionRankingLeaderboardTest))
    return suite


leaderboard/competition_ranking_leaderboard.py
from .leaderboard import Leaderboard
from redis import StrictRedis, Redis, ConnectionPool
import math


class CompetitionRankingLeaderboard(Leaderboard):

    def rank_for_in(self, leaderboard_name, member):
        '''
        Retrieve the rank for a member in the named leaderboard.

        @param leaderboard_name [String] Name of the leaderboard.
        @param member [String] Member name.
        @return the rank for a member in the leaderboard.
        '''
        member_score = str(float(self.score_for_in(leaderboard_name, member)))
        if self.order == self.ASC:
            try:
                return self.redis_connection.zcount(
                    leaderboard_name, '-inf', '(%s' % member_score) + 1
            except:
                return None
        else:
            try:
                return self.redis_connection.zcount(
                    leaderboard_name, '(%s' % member_score, '+inf') + 1
            except:
                return None

    def score_and_rank_for_in(self, leaderboard_name, member):
        '''
        Retrieve the score and rank for a member in the named leaderboard.

        @param leaderboard_name [String]Name of the leaderboard.
        @param member [String] Member name.
        @return the score and rank for a member in the named leaderboard as a Hash.
        '''
        pipeline = self.redis_connection.pipeline()
        pipeline.zscore(leaderboard_name, member)
        if self.order == self.ASC:
            pipeline.zrank(leaderboard_name, member)
        else:
            pipeline.zrevrank(leaderboard_name, member)
        responses = pipeline.execute()

        if responses[0] is not None:
            responses[0] = float(responses[0])

        if self.order == self.ASC:
            try:
                responses[1] = self.redis_connection.zcount(
                    leaderboard_name, '-inf', "(%s" % str(float(responses[0]))) + 1
            except:
                responses[1] = None
        else:
            try:
                responses[1] = self.redis_connection.zcount(
                    leaderboard_name, "(%s" % str(float(responses[0])), '+inf') + 1
            except:
                responses[1] = None

        return {
            self.MEMBER_KEY: member,
            self.SCORE_KEY: responses[0],
            self.RANK_KEY: responses[1]
        }

    def ranked_in_list_in(self, leaderboard_name, members, **options):
        '''
        Retrieve a page of leaders from the named leaderboard for a given list of members.

        @param leaderboard_name [String] Name of the leaderboard.
        @param members [Array] Member names.
        @param options [Hash] Options to be used when retrieving the page from the named leaderboard.
        @return a page of leaders from the named leaderboard for a given list of members.
        '''
        ranks_for_members = []
        scores = []

        pipeline = self.redis_connection.pipeline()

        for member in members:
            if self.order == self.ASC:
                pipeline.zrank(leaderboard_name, member)
            else:
                pipeline.zrevrank(leaderboard_name, member)

            pipeline.zscore(leaderboard_name, member)

        responses = pipeline.execute()

        for index, member in enumerate(members):
            data = {}
            data[self.MEMBER_KEY] = member

            score = responses[index * 2 + 1]
            if score is not None:
                score = float(score)
            else:
                if not options.get('include_missing', True):
                    continue

            data[self.SCORE_KEY] = score

            ranks_for_members.append(data)
            scores.append(data[self.SCORE_KEY])

        for index, rank in enumerate(self.__rankings_for_members_having_scores_in(leaderboard_name, members, scores)):
            ranks_for_members[index][self.RANK_KEY] = rank

        if ('with_member_data' in options) and (True == options['with_member_data']):
            for index, member_data in enumerate(self.members_data_for_in(leaderboard_name, members)):
                ranks_for_members[index][self.MEMBER_DATA_KEY] = member_data

        if 'sort_by' in options:
            if self.RANK_KEY == options['sort_by']:
                ranks_for_members = sorted(
                    ranks_for_members,
                    key=lambda member: member[
                        self.RANK_KEY])
            elif self.SCORE_KEY == options['sort_by']:
                ranks_for_members = sorted(
                    ranks_for_members,
                    key=lambda member: member[
                        self.SCORE_KEY])

        return ranks_for_members

    def __up_rank(self, rank):
        if rank is not None:
            return rank + 1
        else:
            return None

    def __rankings_for_members_having_scores_in(self, leaderboard_name, members, scores):
        pipeline = self.redis_connection.pipeline()

        for index, member in enumerate(members):
            if self.order == self.ASC:
                try:
                    pipeline.zcount(leaderboard_name, '-inf', "(%s" % str(float(scores[index])))
                except:
                    None
            else:
                try:
                    pipeline.zcount(leaderboard_name, "(%s" % str(float(scores[index])), '+inf')
                except:
                    None

        responses = pipeline.execute()

        return [self.__up_rank(response) for response in responses]


leaderboard/tie_ranking_leaderboard.py
from .leaderboard import Leaderboard
from .leaderboard import grouper
from redis import StrictRedis, Redis, ConnectionPool
import math


class TieRankingLeaderboard(Leaderboard):
    DEFAULT_TIES_NAMESPACE = 'ties'

    def __init__(self, leaderboard_name, **options):
        '''
        Initialize a connection to a specific leaderboard. By default, will use a
        redis connection pool for any unique host:port:db pairing.

        The options and their default values (if any) are:

        host : the host to connect to if creating a new handle ('localhost')
        port : the port to connect to if creating a new handle (6379)
        db : the redis database to connect to if creating a new handle (0)
        page_size : the default number of items to return in each page (25)
        connection : an existing redis handle if re-using for this leaderboard
        connection_pool : redis connection pool to use if creating a new handle
        '''
        self.options = options
        self.ties_namespace = self.options.pop(
            'ties_namespace',
            self.DEFAULT_TIES_NAMESPACE)
        self.leaderboard_name = leaderboard_name

        super(TieRankingLeaderboard, self).__init__(
            leaderboard_name, **options)

    def delete_leaderboard_named(self, leaderboard_name):
        '''
        Delete the named leaderboard.

        @param leaderboard_name [String] Name of the leaderboard.
        '''
        pipeline = self.redis_connection.pipeline()
        pipeline.delete(leaderboard_name)
        pipeline.delete(self._member_data_key(leaderboard_name))
        pipeline.delete(self._ties_leaderboard_key(leaderboard_name))
        pipeline.execute()

    def change_score_for_member_in(self, leaderboard_name, member, delta, member_data=None):
        '''
        Change the score for a member in the named leaderboard by a delta which can be positive or negative.

        @param leaderboard_name [String] Name of the leaderboard.
        @param member [String] Member name.
        @param delta [float] Score change.
        @param member_data [String] Optional member data.
        '''
        previous_score = self.score_for(member)
        new_score = (previous_score or 0) + delta

        total_members_at_previous_score = []
        if previous_score is not None:
            total_members_at_previous_score = self.redis_connection.zrevrangebyscore(leaderboard_name, previous_score, previous_score)

        pipeline = self.redis_connection.pipeline()
        if isinstance(self.redis_connection, Redis):
            pipeline.zadd(leaderboard_name, member, new_score)
            pipeline.zadd(self._ties_leaderboard_key(leaderboard_name), str(float(new_score)), new_score)
        else:
            pipeline.zadd(leaderboard_name, new_score, member)
            pipeline.zadd(self._ties_leaderboard_key(leaderboard_name), new_score, str(float(new_score)))
        if member_data:
            pipeline.hset(
                self._member_data_key(leaderboard_name),
                member,
                member_data)
        pipeline.execute()

        if len(total_members_at_previous_score) == 1:
            self.redis_connection.zrem(self._ties_leaderboard_key(leaderboard_name), str(float(previous_score)))

    def rank_member_in(
            self, leaderboard_name, member, score, member_data=None):
        '''
        Rank a member in the named leaderboard.

        @param leaderboard_name [String] Name of the leaderboard.
        @param member [String] Member name.
        @param score [float] Member score.
        @param member_data [String] Optional member data.
        '''
        member_score = None or self.redis_connection.zscore(leaderboard_name, member)
        can_delete_score = member_score is not None and\
            (len(self.members_from_score_range_in(leaderboard_name, member_score, member_score)) == 1) and\
            member_score != score

        pipeline = self.redis_connection.pipeline()
        if isinstance(self.redis_connection, Redis):
            pipeline.zadd(leaderboard_name, member, score)
            pipeline.zadd(self._ties_leaderboard_key(leaderboard_name),
                          str(float(score)), score)
        else:
            pipeline.zadd(leaderboard_name, score, member)
            pipeline.zadd(self._ties_leaderboard_key(leaderboard_name),
                          score, str(float(score)))
        if can_delete_score:
            pipeline.zrem(self._ties_leaderboard_key(leaderboard_name),
                          str(float(member_score)))
        if member_data:
            pipeline.hset(
                self._member_data_key(leaderboard_name),
                member,
                member_data)
        pipeline.execute()

    def rank_member_across(
            self, leaderboards, member, score, member_data=None):
        '''
        Rank a member across multiple leaderboards.

        @param leaderboards [Array] Leaderboard names.
        @param member [String] Member name.
        @param score [float] Member score.
        @param member_data [String] Optional member data.
        '''
        for leaderboard_name in leaderboards:
            self.rank_member_in(leaderboard, member, score, member_data)

    def rank_members_in(self, leaderboard_name, members_and_scores):
        '''
        Rank an array of members in the named leaderboard.

        @param leaderboard_name [String] Name of the leaderboard.
        @param members_and_scores [Array] Variable list of members and scores.
        '''
        for member, score in grouper(2, members_and_scores):
            self.rank_member_in(leaderboard_name, member, score)

    def remove_member_from(self, leaderboard_name, member):
        '''
        Remove the optional member data for a given member in the named leaderboard.

        @param leaderboard_name [String] Name of the leaderboard.
        @param member [String] Member name.
        '''
        member_score = None or self.redis_connection.zscore(
            leaderboard_name, member)
        can_delete_score = member_score and len(
            self.members_from_score_range_in(leaderboard_name, member_score, member_score)) == 1

        pipeline = self.redis_connection.pipeline()
        pipeline.zrem(leaderboard_name, member)
        if can_delete_score:
            pipeline.zrem(self._ties_leaderboard_key(leaderboard_name),
                          str(float(member_score)))
        pipeline.hdel(self._member_data_key(leaderboard_name), member)
        pipeline.execute()

    def rank_for_in(self, leaderboard_name, member):
        '''
        Retrieve the rank for a member in the named leaderboard.

        @param leaderboard_name [String] Name of the leaderboard.
        @param member [String] Member name.
        @return the rank for a member in the leaderboard.
        '''
        member_score = self.score_for_in(leaderboard_name, member)
        if self.order == self.ASC:
            try:
                return self.redis_connection.zrank(
                    self._ties_leaderboard_key(leaderboard_name), str(float(member_score))) + 1
            except:
                return None
        else:
            try:
                return self.redis_connection.zrevrank(
                    self._ties_leaderboard_key(leaderboard_name), str(float(member_score))) + 1
            except:
                return None

    def remove_members_in_score_range_in(
            self, leaderboard_name, min_score, max_score):
        '''
        Remove members from the named leaderboard in a given score range.

        @param leaderboard_name [String] Name of the leaderboard.
        @param min_score [float] Minimum score.
        @param max_score [float] Maximum score.
        '''
        pipeline = self.redis_connection.pipeline()
        pipeline.zremrangebyscore(
            leaderboard_name,
            min_score,
            max_score)
        pipeline.zremrangebyscore(
            self._ties_leaderboard_key(leaderboard_name),
            min_score,
            max_score)
        pipeline.execute()

    def expire_leaderboard_for(self, leaderboard_name, seconds):
        '''
        Expire the given leaderboard in a set number of seconds. Do not use this with
        leaderboards that utilize member data as there is no facility to cascade the
        expiration out to the keys for the member data.

        @param leaderboard_name [String] Name of the leaderboard.
        @param seconds [int] Number of seconds after which the leaderboard will be expired.
        '''
        pipeline = self.redis_connection.pipeline()
        pipeline.expire(leaderboard_name, seconds)
        pipeline.expire(self._ties_leaderboard_key(leaderboard_name), seconds)
        pipeline.expire(self._member_data_key(leaderboard_name), seconds)
        pipeline.execute()

    def expire_leaderboard_at_for(self, leaderboard_name, timestamp):
        '''
        Expire the given leaderboard at a specific UNIX timestamp. Do not use this with
        leaderboards that utilize member data as there is no facility to cascade the
        expiration out to the keys for the member data.

        @param leaderboard_name [String] Name of the leaderboard.
        @param timestamp [int] UNIX timestamp at which the leaderboard will be expired.
        '''
        pipeline = self.redis_connection.pipeline()
        pipeline.expireat(leaderboard_name, timestamp)
        pipeline.expireat(
            self._ties_leaderboard_key(leaderboard_name), timestamp)
        pipeline.expireat(self._member_data_key(leaderboard_name), timestamp)
        pipeline.execute()

    def ranked_in_list_in(self, leaderboard_name, members, **options):
        '''
        Retrieve a page of leaders from the named leaderboard for a given list of members.

        @param leaderboard_name [String] Name of the leaderboard.
        @param members [Array] Member names.
        @param options [Hash] Options to be used when retrieving the page from the named leaderboard.
        @return a page of leaders from the named leaderboard for a given list of members.
        '''
        ranks_for_members = []

        pipeline = self.redis_connection.pipeline()

        for member in members:
            if self.order == self.ASC:
                pipeline.zrank(leaderboard_name, member)
            else:
                pipeline.zrevrank(leaderboard_name, member)

            pipeline.zscore(leaderboard_name, member)

        responses = pipeline.execute()

        for index, member in enumerate(members):
            data = {}
            data[self.MEMBER_KEY] = member

            score = responses[index * 2 + 1]
            if score is not None:
                score = float(score)
            data[self.SCORE_KEY] = score

            if self.order == self.ASC:
                data[self.RANK_KEY] = self.redis_connection.zrank(
                    self._ties_leaderboard_key(leaderboard_name), str(data[self.SCORE_KEY]))
            else:
                data[self.RANK_KEY] = self.redis_connection.zrevrank(
                    self._ties_leaderboard_key(leaderboard_name), str(data[self.SCORE_KEY]))
            if data[self.RANK_KEY] is not None:
                data[self.RANK_KEY] += 1
            else:
                if not options.get('include_missing', True):
                    continue

            ranks_for_members.append(data)

        if ('with_member_data' in options) and (True == options['with_member_data']):
            for index, member_data in enumerate(self.members_data_for_in(leaderboard_name, members)):
                try:
                    ranks_for_members[index][self.MEMBER_DATA_KEY] = member_data
                except:
                    pass

        if 'sort_by' in options:
            if self.RANK_KEY == options['sort_by']:
                ranks_for_members = sorted(
                    ranks_for_members,
                    key=lambda member: member[
                        self.RANK_KEY])
            elif self.SCORE_KEY == options['sort_by']:
                ranks_for_members = sorted(
                    ranks_for_members,
                    key=lambda member: member[
                        self.SCORE_KEY])

        return ranks_for_members

    def _ties_leaderboard_key(self, leaderboard_name):
        '''
        Key for ties leaderboard.

        @param leaderboard_name [String] Name of the leaderboard.
        @return a key in the form of +leaderboard_name:ties_namespace+
        '''
        return '%s:%s' % (leaderboard_name, self.ties_namespace)
